pymatgen/analysis/diffraction/tem.py

```diff
@@ -0,0 +1,520 @@
+        # second is the first non-parallel-to-first-point vector when sorted.
+        # note 000 is "parallel" to every plane vector.
+        for plane in sorted(spacings.keys()):
+            second_point, second_d = plane, spacings[plane]
+            if not self.is_parallel(first_point, second_point):
+                break
+        p1 = first_point
+        p2 = second_point
+        if (0, 0, 0) in points:
+            points.remove((0, 0, 0))
+        points.remove(first_point)
+        points.remove(second_point)
+        positions[(0, 0, 0)] = np.array([0, 0])
+        r1 = 10 ** 10 * self.wavelength_rel() * self.camera_length / first_d
+        positions[first_point] = np.array([r1, 0])
+        r2 = 10 ** 10 * self.wavelength_rel() * self.camera_length / second_d
+        phi = np.arccos(
+            np.dot(p1, np.transpose(p2)) / (np.sqrt(np.dot(p1, np.transpose(p1)) * np.dot(p2, np.transpose(p2)))))
+        positions[second_point] = np.array([r2 * np.cos(phi), r2 * np.sin(phi)])
+        denom = np.linalg.det(np.array([[p1[0], p1[1]], [p2[0], p2[1]]]))
+        init_denom_0 = (denom == 0)
+        if init_denom_0:
+            denom = np.linalg.det(np.array([[p1[0], p1[2]], [p2[0], p2[2]]]))
+        for plane in points:
+            coeffs = self.get_plot_coeffs(p1, p2, plane, denom, init_denom_0)
+            pos = np.array([coeffs[0] * positions[first_point][0] + coeffs[1] * positions[second_point][0],
+                            coeffs[0] * positions[first_point][1] + coeffs[1] * positions[second_point][1]])
+            positions[plane] = pos
+        points.append((0, 0, 0))
+        points.append(first_point)
+        points.append(second_point)
+
+        return positions
+
+    def TEM_dots(self, structure: Structure, points: list) -> list:
+        """
+        Generates all TEM_dot as named tuples that will appear on the 2D diffraction pattern.
+        Args:
+            structure (Structure): The input structure.
+            points (list): All points to be checked.
+        Returns:
+            list of TEM_dots
+        """
+        dots = []
+        interplanar_spacings = self.get_interplanar_spacings(structure, points)
+        bragg_angles = self.bragg_angles(interplanar_spacings)
+        cell_intensity = self.normalized_cell_intensity(structure, bragg_angles)
+        positions = self.get_positions(structure, points)
+        for plane in cell_intensity.keys():
+            dot = namedtuple('TEM_dot', ['position', 'hkl', 'intensity', 'film_radius', 'd_spacing'])
+            position = positions[plane]
+            hkl = plane
+            intensity = cell_intensity[plane]
+            film_radius = 0.91 * (10 ** -3 * self.cs * self.wavelength_rel() ** 3) ** Fraction('1/4')
+            d_spacing = interplanar_spacings[plane]
+            TEM_dot = dot(position, hkl, intensity, film_radius, d_spacing)
+            dots.append(TEM_dot)
+        return dots
+
+    def show_plot_2d(self, structure: Structure):
```

Comment:
Method is a lot more useful if it returns the `fig` object directly

Reply:
Fixed!